### PR TITLE
Add "icmp_code" to whitelist of allowed rule fields.

### PR DESCRIPTION
**Pull request by fasaxc**
Also beef up UTs and validation to check whitelist against rules.

This fixes the issue at hand but I'm not sure we can merge without addressing #451 in some way.

### CODE REVIEW

**Comment by Lukasa** [comment]
This is not assigned to a person or a release, which means it'll languish.

**Comment by fasaxc** [comment]
Hmm, it was meant to be with me or Rob, I think.  Github was broken last night; losing comments and updates.  (I was also tired so may have messed up.)

**Comment by fasaxc** [comment]
@robbrockbank found in manual testing that the ICMP type/code match of 255/255 is interpreted by the kernel as a wildcard so I've added a workaround to this fix.  (It's not safe to merge the fix without the workaround because it would open up a security hole.)

Resolving the kernel issue is tracked under #451.

**Comment by fasaxc** [comment]
(Workaround is to log an error and generate an unconditional DROP rule if someone tries to use 255/255.)

File: calico/felix/frules.py [hunk 1]
```diff
@@ -53,6 +53,7 @@
     "dst_tag",
     "dst_ports",
     "icmp_type",
+    "icmp_code",
```
Comment (fasaxc) [inline calico/felix/frules.py]: Actual bug fix here.

File: calico/openstack/test/test_plugin_etcd.py [hunk 2]
```diff
@@ -551,10 +551,10 @@ def assertNeutronToEtcd(self, neutron_rule, exp_etcd_rule):
 
         # Check felix is happy with generated rule.
         if neutron_rule["direction"] == "ingress":
-            rules = {"inbound_rules": [neutron_rule],
+            rules = {"inbound_rules": [etcd_rule],
```
Comment (fasaxc) [inline calico/openstack/test/test_plugin_etcd.py]: UT bug: was validating wrong dict!

File: calico/felix/frules.py [hunk 3]
```diff
@@ -199,16 +194,22 @@ def rule_to_iptables_fragments(chain_name, rule, ip_version, tag_to_ipset,
...
+            rule_copy["src_ports"] = src_ports
+            rule_copy["dst_ports"] = dst_ports
+            frag = _rule_to_iptables_fragment(chain_name, rule_copy, ip_version,
+                                              tag_to_ipset,  on_allow=on_allow,
+                                              on_deny=on_deny)
+            fragments.append(frag)
+        return fragments
+    except Exception:
+        # Defensive: isolate failures to parse the rule (which has already
+        # passed validation byt his point) to this chain.
```
Comment (robbrockbank) [inline calico/felix/frules.py]: typo:
  byt his

**Comment by robbrockbank** [comment]
LGTM

**Pull request merged**